jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,617 @@
+ *  - tk_2 = 10005, price is $1.05 ($1.00 + 5 penny)
+ * - rel_X : size_t type. A relative position of a price at top_levels_
+ *  - BUY side examples:
+ *   - rel_1 = 20 (with px_begin_top_ = $30.00), tk_1 = 12920, px_1 = $30.20
+ *   - rel_2 = 40 (with px_begin_top_ = c30.00), tk_2 = 3040, px_2 = c30.40
+ *   - rel_3 = 1020 (with px_begin_top_ = c90.00), tk_3 = 10020, px_3 = $1.20
+ *  - SELL side examples :
+ *   - rel_1 = 20 (with px_begin_top_ = $30.00), tk_1 = 12880, px_1 = $29.80
+ *   - rel_2 = 40 (with px_begin_top_ = c30.00), tk_2 = 2960, px_2 = c29.60
+ *   - rel_3 = 1020 (with px_begin_top_ = $1.10), tk_3 = 8990, px_3 = c89.90
+ */
+template <typename compare_t>
+class array_based_book_side {
+public:
+  explicit array_based_book_side(std::size_t sz)
+      : better_()
+      , max_size_(sz)
+      , top_levels_(sz, 0)
+      , bottom_levels_()
+      , px_inside_(empty_quote().first.as_integer())
+      , px_begin_top_(empty_quote().first.as_integer())
+      , px_end_top_(empty_quote().first.as_integer()) {
+  }
+
+  /// The value used to represent an empty bid
+  static half_quote empty_bid() {
+    return half_quote(LOWEST_PRICE, 0);
+  }
+
+  /// The value used to represent an empty offer
+  static half_quote empty_offer() {
+    return half_quote(HIGHEST_PRICE, 0);
+  }
+
+  /// @returns an empty bid or offer based on compare function
+  /// empty bid for less, empty offer for greater.
+  half_quote empty_quote() const {
+    if (better_(price4_t(1), price4_t(0))) {
+      // buy side
+      return empty_bid();
+    }
+    // sell side
+    return empty_offer();
+  }
+
+  /// @returns the best bid price and quantity.
+  /// The inside is always at the top_levels_ (empty_quote.price when empty)
+  half_quote best_quote() const {
+    if (px_inside_ == empty_quote().first) {
+      return empty_quote();
+    }
+    int rel_px = price_to_relative(px_inside_);
+    return half_quote(px_inside_, top_levels_.at(rel_px));
+  }
+
+  /// @returns the worst bid price and quantity.
+  /// Check if the side is empty first (return empty_quote in this case)
+  /// The worst price is at the bottom_levels (if not empty),
+  /// at the top_levels otherwise.
+  half_quote worst_quote() const {
+    if (px_inside_ == empty_quote().first) {
+      return empty_quote(); // empty side
+    }
+    if (!bottom_levels_.empty()) {
+      // ... worst price is at the bottom_levels_
+      auto i = bottom_levels_.rbegin(); // get the worst bottom_levels price
+      return half_quote(i->first, i->second);
+    }
+    // ... worst price at the top_levels_
+    int rel_worst = relative_worst_top_level();
+    price4_t px_worst = relative_to_price(rel_worst);
+    return half_quote(px_worst, top_levels_.at(rel_worst));
+  }
+
+  /// @returns the number of levels with non-zero quantity for the order side.
+  /// This is the size of bottom_levels_ plus all non zero top_levels values
+  std::size_t count() const {
+    if (!bottom_levels_.empty()) {
```

Comment:
As a matter of style, I have adopted the C++11 boolean operators 'not', 'and', 'or', etc.  So that expression would become:

    if (not bottom_levels_.empty()) {

second, if `bottom_levels_.empty()` is true, then `bottom_levels_.size()` returns 0, so the if() is not really needed?  I thought the time complexity for `bottom_levels_.size()` is constant, or not?